### PR TITLE
use lossless conversions where available

### PR DESCRIPTION
**Pull request by danieleades**

### CODE REVIEW

**Comment by tertsdiepraam** [comment]
I tried the conversion from `u64` to `u128` in Godbolt and found that many of these compiled to the same thing, but calling `From` might have the added overhead of a function call. So what is the advantage here? Aren't both options lossless?

**Comment by danieleades** [comment]
> I tried the conversion from `u64` to `u128` in Godbolt and found that many of these compiled to the same thing, but calling `From` might have the added overhead of a function call. So what is the advantage here? Aren't both options lossless?
> 
> Here is the code I put in Godbolt: https://godbolt.org/z/sbrY5EsMW

to be honest i'm pretty surprised the compiler doesn't optimise away the difference...

the main benefit of using the conversion methods is safety. If one of the types were to change at any point a cast could become lossy, an explicit conversion method cannot.

choose your compromise, i guess

File: src/uucore/src/lib/features/utmpx.rs [hunk 1]
```diff
@@ -191,7 +191,7 @@ impl Utmpx {
     /// A.K.A. ut.ut_tv
     pub fn login_time(&self) -> Tm {
         time::at(Timespec::new(
-            self.inner.ut_tv.tv_sec as i64,
+            self.inner.ut_tv.tv_sec.into(),
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/utmpx.rs:194]: You did trigger a lint lint that actually also applies to the original code

File: src/uucore/src/lib/parser/parse_time.rs [hunk 2]
```diff
@@ -37,7 +37,7 @@ pub fn from_str(string: &str) -> Result<Duration, String> {
 
     const NANOS_PER_SEC: u32 = 1_000_000_000;
     let whole_secs = num.trunc();
-    let nanos = (num.fract() * (NANOS_PER_SEC as f64)).trunc();
+    let nanos = (num.fract() * f64::from(NANOS_PER_SEC)).trunc();
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/parser/parse_time.rs:40]: Here, the constant probably should have been `f64` from the start

**Review by tertsdiepraam — COMMENTED** [review]
> to be honest i'm pretty surprised the compiler doesn't optimise away the difference...

Me too actually. Might be an issue with Godbolt to be fair. I'm not sure what profile it uses. 

> If one of the types were to change at any point a cast could become lossy, an explicit conversion method cannot.

I find it very unlikely that `u64` will at some point no longer be a 64-bit unsigned integer :). For more complex types I would accept this change, but for numbers I think it's fine like it is.

@sylvestre what's your take on this?

**Comment by tertsdiepraam** [comment]
Small update: I tried `cargo asm` instead of Godbolt and both methods were completely optimized away indeed:
```rust
pub fn a(a: u64, b: u64) -> u128 {
    let a2 = u128::from(a);
    let b2 = b as u128;
    a2 + b2
}
```
```
❯ cargo asm test::a --rust
 pub fn a(a: u64, b: u64) -> u128 {
 mov     rax, rdi
 a2 + b2
 xor     edx, edx
 add     rax, rsi
 setb    dl
 }
 ret
```

File: src/uucore/src/lib/parser/parse_time.rs (hunk 2 above)
Comment (danieleades) [reply src/uucore/src/lib/parser/parse_time.rs]: good catch! updated.

File: src/uucore/src/lib/features/utmpx.rs (hunk 1 above)
Comment (danieleades) [reply src/uucore/src/lib/features/utmpx.rs:194]: what do you mean?

Comment (tertsdiepraam) [reply src/uucore/src/lib/features/utmpx.rs:194]: Clippy is saying that it's a useless conversion on Mac.

**Comment by miDeb** [comment]
> Me too actually. Might be an issue with Godbolt to be fair. I'm not sure what profile it uses.

That indeed seems to be the issue, if you add `-C opt-level=3` to the compiler options it will optimize it correctly (it will even remove one function).

The reason why clippy prefers `.into()` is that `as` _may_ do a lossy conversion. `.into()` won't, so if the types were to change, and the conversion couldn't be guaranteed to be lossless anymore, it would no longer compile.
At least that's my understanding.

**Comment by danieleades** [comment]
hmm this is going to be a little subtle to solve, as the types vary between platforms.

**Comment by tertsdiepraam** [comment]
In those cases it's okay to just use `into()` and put an `allow` on it. So this one: 
```
tv_sec: date.timestamp(),
        ^^^^^^^^^^^^^^^^ expected `i32`, found `i64`
```
Could be
```rust
#[allow(clippy::useless_conversion)]
tv_sec: date.timestamp().into(),
```

For this one, I think we can say that it will never exceed `2^31-1` because that's still greater than `10^9` (the number of nanoseconds in a second). So we can suppress the warning and use "lossy" conversion.
```
tv_nsec: date.timestamp_subsec_nanos().into(),
                                       ^^^^ the trait `From<u32>` is not implemented for `i32`
```

**Comment by sylvestre** [comment]
@danieleades ping?

**Comment by sylvestre** [comment]
No action for a while, closing. Please reopen when ready!

**Pull request closed without merging**